File: framework/languages/java/samples/java/GameQuest/Server/GameApi/src/main/java/systems/zlink/samples/gamequest/server/gameapi/domain/GameplayDomain.java
package systems.zlink.samples.gamequest.server.gameapi.domain;

import systems.zlink.samples.gamequest.shared.contracts.Messages;

/** Factory for gameplay event envelopes. Mirrors the .NET {@code GameplayDomain}. */
public final class GameplayDomain {
    private GameplayDomain() {
    }

    public static Messages.GameplayEventEnvelope createMonsterKilled(
        String playerId, String monsterId, String areaId, String idempotencyKey, String sourceApi) {
        return create(playerId, idempotencyKey, "MonsterKilled", monsterId, 1, sourceApi);
    }

    public static Messages.GameplayEventEnvelope createItemCollected(
        String playerId, String itemId, int count, String idempotencyKey, String sourceApi) {
        return create(playerId, idempotencyKey, "ItemCollected", itemId, count, sourceApi);
    }

    public static Messages.GameplayEventEnvelope createMissionCompleted(
        String playerId, String missionId, String idempotencyKey, String sourceApi) {
        return create(playerId, idempotencyKey, "MissionCompleted", missionId, 1, sourceApi);
    }

    public static Messages.GameplayEventEnvelope createAreaEntered(
        String playerId, String areaId, String idempotencyKey, String sourceApi) {
        return create(playerId, idempotencyKey, "AreaEntered", areaId, 1, sourceApi);
    }

    public static Messages.GameplayEventEnvelope createFeatureUnlocked(
        String playerId, String featureId, String idempotencyKey, String sourceApi) {
        return create(playerId, idempotencyKey, "FeatureUnlocked", featureId, 1, sourceApi);
    }

    private static Messages.GameplayEventEnvelope create(
        String playerId, String idempotencyKey, String eventType, String value, int count, String sourceApi) {
        if (playerId == null || playerId.isBlank()) {
            throw new IllegalStateException("Player id is required.");
        }
        if (idempotencyKey == null || idempotencyKey.isBlank()) {
            throw new IllegalStateException("Idempotency key is required.");
        }
        if (count <= 0) {
            throw new IllegalStateException("Count must be positive.");
        }
        return new Messages.GameplayEventEnvelope(
            playerId + "-" + idempotencyKey,
            playerId,
            idempotencyKey,
            eventType,
            value,
            count,
            sourceApi,
            System.currentTimeMillis());
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/GameApi/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/gameapi/domain/GameplayDomain.kt
package systems.zlink.samples.kotlin.gamequest.server.gameapi.domain

import systems.zlink.samples.kotlin.gamequest.shared.contracts.GameplayEventMsg

/** Factory for gameplay event envelopes. Mirrors the .NET `GameplayDomain`. */
object GameplayDomain {
    fun createMonsterKilled(
        playerId: String,
        monsterId: String,
        areaId: String,
        idempotencyKey: String,
        sourceApi: String,
    ): GameplayEventMsg =
        create(playerId, idempotencyKey, "MonsterKilled", monsterId, 1, sourceApi)

    fun createItemCollected(
        playerId: String,
        itemId: String,
        count: Int,
        idempotencyKey: String,
        sourceApi: String,
    ): GameplayEventMsg =
        create(playerId, idempotencyKey, "ItemCollected", itemId, count, sourceApi)

    fun createMissionCompleted(
        playerId: String,
        missionId: String,
        idempotencyKey: String,
        sourceApi: String,
    ): GameplayEventMsg =
        create(playerId, idempotencyKey, "MissionCompleted", missionId, 1, sourceApi)

    fun createAreaEntered(
        playerId: String,
        areaId: String,
        idempotencyKey: String,
        sourceApi: String,
    ): GameplayEventMsg =
        create(playerId, idempotencyKey, "AreaEntered", areaId, 1, sourceApi)

    fun createFeatureUnlocked(
        playerId: String,
        featureId: String,
        idempotencyKey: String,
        sourceApi: String,
    ): GameplayEventMsg =
        create(playerId, idempotencyKey, "FeatureUnlocked", featureId, 1, sourceApi)

    private fun create(
        playerId: String,
        idempotencyKey: String,
        eventType: String,
        value: String,
        count: Int,
        sourceApi: String,
    ): GameplayEventMsg {
        check(playerId.isNotBlank()) { "Player id is required." }
        check(idempotencyKey.isNotBlank()) { "Idempotency key is required." }
        check(count > 0) { "Count must be positive." }
        return GameplayEventMsg(
            "$playerId-$idempotencyKey",
            playerId,
            idempotencyKey,
            eventType,
            value,
            count,
            sourceApi,
            System.currentTimeMillis(),
        )
    }
}
